### ISSUE TITLE
Loss of data conversion in Flex scanner.

### ISSUE THREAD

**Issue by papaathome** [lexxmark/winflexbison#73]
In a test project I get the following warning: Details will follow below.

warning C4244: 'return': conversion from 'std::streamsize' to 'int', possible loss of data

Related code:
`// YY_INTERACTIVE is not defined, refactored (by removing preprocessor directives) to make it shorter.`
`int yyFlexLexer::LexerInput( char* buf, int max_size )`
`{`
`    if ( yyin.eof() || yyin.fail() ) return 0;`
`    (void) yyin.read( buf, max_size );`
`    if ( yyin.bad() ) return -1;`
`    else return yyin.gcount(); // <-- This line is causing the warning`
`}`

I am using Visual Studio 2017 with the latest version of win_bison_flex (win_flex_bison3-latest.zip)
As test project I use the source from [Flex Bison C++ Example](https://panthema.net/2007/flex-bison-cpp-example/flex-bison-cpp-example-0.1.4/doxygen-html/index.html) 
There were a few minor things I had to fix but the code compiles and runs without apparent problems.

The scanner code generated by Flex is causing this warning.
I do not see a way to redefine or otherwise fix the definition of `int yyFlexLexer::LexerInput(...)`.
I would like to change it to `long long yyFlexLexer::LexerInput(...)`

Does anybody have an advise on how to handle this problem?
When parsing large files this can become an issue and I don't like compiler warnings anyway.

Kind regards.

**Comment by lexxmark** [comment]
Could you please check that upstream bison doesn't produce such warning (under linux)?

If it doesn't produce a warning I should investigate that's the difference, otherwise please report that problem to upstream [bison ](https://github.com/akimd/bison) project to fix.

**Comment by papaathome** [comment]
Sorry, I don't have a *nix development environment available at the moment.

It will take me some time to get one up (using VirtualBox) to do a quick test.
But when I find the time for it I will report all details to upstream bison and mention the results here.

Kind regards,
Andre.

**Comment by ArcaneTourist** [comment]
LexerInput() is a virtual function, right?   So, you must have provided the code.

The return value for LexerInput is an int from 0..max_size.   The way you coded your call to read() guarantees that the value from gcount() must be less than max_size and thus, must fit in an int.  So, you should cast the result of gcount() to int and return an int as expected by the flex interface conventions.

This issue is due to the user defined code and should be closed.

**Comment by GitMensch** [comment]
> LexerInput() is a virtual function, right? So, you must have provided the code.

No,that is generated by flex or in this case win_flex into scanner.cc and also prototyped in [FlexLexer.h](https://github.com/lexxmark/winflexbison/blob/159682c7b49f858d83ea57736a045f2d54316c58/flex/src/FlexLexer.h#L148).

The [referenced sample](https://panthema.net/2007/flex-bison-cpp-example/flex-bison-cpp-example-0.1.4.tar.bz2) has those pre-generated / shipped and back then it all was of type `int`, too.

Current flex [still has that as an int](https://github.com/westes/flex/blob/2995af0eeb3792506f2b9975788443f67ac4b389/src/FlexLexer.h#L149), and the [current cpp skeleton](https://github.com/westes/flex/blob/7b073fdda2242a8e4a0f70244c0ae83d34f099b1/src/cpp-flex.skl#L2393) also returns `yyin.gcount()`. So we are "identical to upstream.

There is currently no actual issue as the stream reads up to the max_size, which is also an int, the implementation would be completely broken if it returns anything bigger.

Depending on the actual environment [`std::istream::gcount()`](https://www.cplusplus.com/reference/istream/istream/gcount/) is defined differently because it is of type [`streamsize`](https://www.cplusplus.com/reference/ios/streamsize/), [MSVC defines that](https://docs.microsoft.com/en-us/cpp/standard-library/ios-typedefs?view=msvc-160#streamsize) depending on the compilation of 32/64 bit:

```c
#ifdef _WIN64
    typedef __int64 streamsize;
#else
    typedef int streamsize;
#endif
```

So the requested change for `long long` would be wrong for winflexbison. 

But MSVC is not the only one that not always uses `int`, so an upstream change is reasonable.
I think the skeleton should be changed to explicit cast to the return code:

```diff
-		return yyin.gcount();
+		return (int)yyin.gcount();
```

Upstream merge-request: https://github.com/westes/flex/pull/495.

**Comment by ArcaneTourist** [comment]
Agreed.   Thanks.

**Comment by lexxmark** [comment]
Fixed on `dev` (77978ed); it ships in the next release, 2.5.26.

The warning is real but the code was safe: `LexerInput` returns `yyin.gcount()`, a `std::streamsize` (64-bit on x64), as `int`. The `read()` above it is bounded by `max_size`, itself an `int`, so the value always fits — it just never said so. Fixed with the same one-line cast upstream flex made in [`b198864a`](https://github.com/westes/flex/commit/b198864a):

```
-		return yyin.gcount();
+		return (int)yyin.gcount();
```

Two notes:

- It lives in the skeleton (`flex/src/flex.skl`, compiled into `flex/src/skel.c`), not in `FlexLexer.h` — `LexerInput` keeps its `int` return, same as upstream, so nothing changes for existing code.
- Only non-interactive scanners (`-B` / `%option never-interactive`, which the flex-bison-cpp-example uses) compile that line at all; the default interactive path takes a different branch and never warned. That is why it looked unreproducible in a plain `%option c++` scanner.

A test compiling a `-B` C++ scanner with C4244 promoted to an error now guards it.
